### PR TITLE
Evaluate binary and Poisson quantities on the log scale

### PR DESCRIPTION
**Pull request by choxos**
# Summary

Second of the v0.2.0 series. Removes the two clipping helpers from the Stan
models and rebuilds the binary and Poisson likelihoods and estimands on the log
scale.

Both helpers were one-sided, and one of them was biased rather than merely
bounded:

* `safe_logit(p)` clamped `p` to `[1e-10, 1 - 1e-10]` before the logit. A
  marginal probability that had already rounded to 0 or 1 on the natural scale
  contributed the clip boundary to the reported log odds ratio.
* `safe_divide(num, denom)` returned `num / max(denom, 1e-10)`, substituting a
  denominator **larger** than the true one. A risk ratio or rate ratio against a
  near-zero comparator was therefore systematically understated, silently.

Working from logarithms removes most of what triggered the clip in the first
place, because a marginal probability is no longer rounded to 0 or 1 before the
contrast is taken. What remains is that a ratio genuinely overflowing double
precision is now `Inf` rather than a large finite surrogate.

Two other behavior changes ride with it, both consequences of the same idea:

* `predict(type = "link")` returns `g(E[g^-1(eta)])` rather than `E[eta]`.
  Differencing two link predictions now reproduces the reported marginal effect,
  which it did not before.
* `bound_probability()` replaces its clamp with a pseudo-count correction. For
  an observed binomial proportion the clamp could only ever bind at exactly 0 or
  1, since `r >= 1` gives `r/n >= 1/n > min_count/n`, so it was a boundary
  correction written as a general clamp. It now says so.

The probit path is also written as `std_normal_lcdf(-eta)` rather than
`std_normal_lccdf(eta)`. Stan's two normal tail functions are not equally well
conditioned: the complement carries 5e-4 relative error at `eta = 8` and
underflows to `-inf` from 8.5, where the true log non-event probability is a
representable -39.2, while `std_normal_lcdf` is exact to 2.5e-16 out to -300.

## Type of Change

- [x] Methodological/statistical change
- [x] Bug fix

## Statistical or User-Facing Impact

Numbers a 0.1.0 user could have seen change, in three places. No function's
arguments change.

| Quantity | Before | Now |
|---|---|---|
| `rr_*`, rate ratios vs a near-zero comparator | understated by the denominator floor | exact log contrast |
| `lor_*` at an extreme predictor | clip boundary | finite true value |
| `predict(type = "link")` | `E[eta]` | `g(E[g^-1(eta)])` |
| `bound_probability()` at a zero-event arm | `0.5 / n` | `0.5 / (n + 1)` |

Arms with events on both sides are unaffected. The identity link is unaffected
by the `type = "link"` change.

## Scope

**Normal models are deliberately not included.** Their generated-quantities
block at the target state entangles this log-scale change with `agd_weight`, the
sample-size-weighted comparator estimand, which is a separate estimand fix. They
move across in that pull request, where both changes to the same block can be
made coherently. There is no functional gap: the link fallback takes `log()` of
the marginal mean, which is exactly the log-link marginal link.

`test-family-numerical-stability.R` therefore asserts the log-scale property for
the binary and Poisson models only, with a comment naming when the list widens.
Its `.stc_hypot` assertion is likewise held back to the pull request that
changes `R/stc.R`.

## Reading the diff

`src/` is 4365 insertions, and none of it needs reading. Two things are mixed
there. The model change is one. The other is a toolchain refresh: the exports
tracked since 0.1.0 were produced by an older rstan, and the generated C++
embeds the stanc version and a source line and column map, so regenerating
rewrites all six models whether or not their source changed.

Measured by regenerating from the **unchanged** 0.1.0 sources on the current
toolchain, that accounts for **1729 insertions and 903 deletions** of the total,
including the whole of the normal models and every `.cc` file, none of which
were edited here.

## Verification

```text
cmdstanr check_syntax, all 6 models          -> OK
pure-R suite      (NOT_CRAN=false)           -> pass 561  fail 0  err 0  skip 11
full Stan suite   (NOT_CRAN=true, cmdstanr)  -> pass 661  fail 0  err 0  skip 0
lintr on changed R files                     -> 0 lints
```

The link test keeps the old expected values in a negative assertion, so a silent
revert to `E[eta]` fails rather than passing unnoticed.

## Documentation

- [ ] README updated if needed (no change needed)
- [x] Help pages or roxygen updated if needed
- [ ] Vignettes updated if needed (documentation pull request)
- [x] NEWS.md updated for user-facing changes

`?mlumr-numerical-guards` becomes `?mlumr-numerical-evaluation`: what is computed
on the log scale, which constraints are structural rather than numerical, and
how to read an extreme draw.

The NAMESPACE reformatting of the `stats` imports into a single call is a
roxygen2 8.1.0 artifact, not a change of imports.

## Checklist

- [x] The change is focused and reviewable
- [x] Relevant tests pass
- [x] No local build artifacts, `.Rcheck` output, tarballs, or caches are committed
- [x] The pull request follows the Code of Conduct

## Summary by CodeRabbit

* **New Features**
  * Added more stable calculations for binary outcomes, normal-link predictions, rates, ratios, and marginal summaries, including extreme values.
  * Added log-scale outputs for binary probabilities and Poisson rates to support reliable interpretation of very small or large results.
  * Added continuity correction for probabilities at the 0% and 100% boundaries while preserving interior probabilities.

* **Behavior Changes**
  * `predict(type = "link")` now reports the link applied to the population-standardized response mean.
  * Marginal summaries no longer clip values to finite bounds and instead use log-scale calculations.

* **Documentation**
  * Added guidance on log-scale numerical evaluation and updated prediction and boundary-probability documentation.

### CODE REVIEW

File: R/predict.R [hunk 1]
```diff
@@ -25,13 +23,26 @@
...
-#' and the two interpretations coincide.
+#' Stan `generated quantities` block computes.
 #'
-#' @return A data frame with predictions. When `type = "link"`, values are
-#'   mean linear predictors computed directly from parameter draws (avoiding
-#'   Jensen's inequality bias).
+#' `type = "link"` reports `g(E[g^{-1}(eta)])`: the fitted link applied to
+#' that same standardized response mean. This is the quantity
+#' [marginal_effects()] contrasts, so differencing two `type = "link"`
+#' predictions reproduces the reported marginal effect. It is computed from
```
[the remark this answers is not in the document]
Comment (choxos) [reply R/predict.R]: Valid, fixed. The claim held only where the fitted-link scale and the reported effect scale coincide, which is a logit binomial fit's `lor_*`. The roxygen and NEWS now say so and name what the other families report instead: Poisson a natural-scale rate ratio (so the link difference is its logarithm), normal log-link a natural-scale mean difference, and probit and cloglog `lor_*`, `rd_*` and `rr_*`, none of which is a fitted-link contrast.

File: R/predict.R [hunk 2]
```diff
@@ -195,84 +204,72 @@ marginal_effects <- function(object,
...
+    link_cols <- paste0("link_y_", suffix)
+    if (all(link_cols %in% names(draws))) {
+      as.data.frame(draws[link_cols])
+    } else {
+      # Older normal-log fits did not store the stable log marginal mean.
+      .require_draw_columns(draws, pred_cols, "marginal link prediction")
+      vals <- as.data.frame(log(draws[pred_cols]))
+      if (any(!is.finite(as.matrix(vals)))) {
+        stop("This older normal-log fit lacks stable marginal link draws; refit ",
+             "the model to obtain them.", call. = FALSE)
```
[the remark this answers is not in the document]
Comment (choxos) [reply R/predict.R:273]: Right, and fixed properly rather than by rewording the error. The normal models now emit `link_y_*` and build their marginal means with `log_mean_exp_vec` / `log_weighted_mean_exp_vec`, including the log-link AgD likelihood, which previously averaged `exp(theta)` on the natural scale.

I had scoped the normal models out because their generated-quantities block at the target state also carries `agd_weight`, a separate estimand change. On a closer look the index-population block is free of it and only the comparator block uses it, so the log-scale work lands here with the 0.1.0 equal weighting preserved, and the weighting change stays with the pull request that makes `outcome_n` required.

File: R/link_functions.R [hunk 3]
```diff
@@ -122,7 +280,9 @@ bound_probability <- function(p, n, min_count = 0.5) {
   if (any(2 * min_count > n)) {
     stop("`min_count` must be no larger than n / 2.", call. = FALSE)
   }
-  pmin(pmax(p, min_count / n), 1 - min_count / n)
+  lower <- min_count / (n + 2 * min_count)
+  upper <- (n + min_count) / (n + 2 * min_count)
+  ifelse(p <= 0, lower, ifelse(p >= 1, upper, p))
```
[the remark this answers is not in the document]
Comment (choxos) [reply R/link_functions.R]: Agreed, fixed. `p` outside [0, 1] is an upstream construction error, so it now errors rather than being mapped to a corrected value; `NA` still passes through; and the pseudo-count applies only at exactly 0 or 1.

File: man/mlumr-numerical-evaluation.Rd [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+
+\describe{
+\item{Binary outcomes}{Event and non-event log probabilities are evaluated
+directly for logit, probit, and complementary log-log links. Aggregate
+probabilities are then formed with \code{log_sum_exp}, so an extreme but
+finite linear predictor is not first rounded to probability 0 or 1.}
+\item{Log-link outcomes}{Marginal normal means and Poisson rates use
+log-mean-exp calculations. Population weights are normalized on the log
+scale, so multiplying every weight by the same finite constant does not
```
[the remark this answers is not in the document]
Comment (choxos) [reply man/mlumr-numerical-evaluation.Rd]: Fixed. Zero and negative constants do not preserve the contract, so the sentence now reads "positive finite constant" in `R/numerical_guards.R` and the .Rd is regenerated.

File: NEWS.md [hunk 5]
```diff
@@ -1,5 +1,51 @@
...
+  ratios are no longer biased downward. The old `safe_divide()` substituted a
+  denominator LARGER than the true one, so a risk ratio or rate ratio with a
+  near-zero comparator was systematically understated; the log-scale contrast
+  reports it. Working on the log scale also removes most of what used to trigger
+  the clip in the first place, because a marginal probability is no longer
+  rounded to 0 or 1 before the contrast is taken: `lor_*` is finite in cases
+  where `safe_logit()` previously returned its clip boundary. What remains is
+  that a ratio whose true value overflows double precision is now `Inf` rather
+  than a large finite surrogate, which needs a draw with an effectively infinite
+  linear predictor. Read the log-scale generated quantities when that happens.
```
[the remark this answers is not in the document]
Comment (choxos) [reply NEWS.md]: Fixed. A finite log contrast above `log(.Machine$double.xmax)`, about 709.78, is enough; NEWS said an effectively infinite linear predictor was needed, which overstated it.

File: R/link_functions.R [hunk 6]
```diff
@@ -94,6 +94,160 @@ inverse_link <- function(x, link = c("identity", "log", "logit", "probit", "clog
...
+
+#' Weighted log mean of exponentiated values
+#' @keywords internal
+.weighted_log_mean_exp <- function(x, weights = rep(1, length(x))) {
+  if (length(x) != length(weights) || any(!is.finite(weights)) ||
+        any(weights < 0) || !any(weights > 0)) {
+    stop("`weights` must be finite, non-negative, and match `x`.", call. = FALSE)
+  }
+  log_weights <- log(weights)
+  z <- x + log_weights
```
[the remark this answers is not in the document]
Comment (choxos) [reply R/link_functions.R:207]: Fixed. Zero-weight entries are dropped before forming `x + log(weights)`, since `Inf + -Inf` is `NaN` and a term contributing nothing was producing a missing result. Regression test added, including that appending a zero-weight entry leaves the result unchanged.

File: R/link_functions.R [hunk 7]
```diff
@@ -94,6 +94,160 @@ inverse_link <- function(x, link = c("identity", "log", "logit", "probit", "clog
...
+  small <- !is.na(eta) & eta < -18
+  if (any(small)) {
+    # 1 - exp(-x) = x (1 - x/2 + x^2/6 - ...), so log(1 - exp(-x)) is
+    # eta + log1p(-x/2 + x^2/6) with x = exp(eta). Below eta = -18 the direct
+    # form loses the leading digits, and below eta = -745 exp(eta) underflows to
+    # zero and it returns -Inf instead of eta.
+    x <- exp_eta[small]
+    log_event[small] <- eta[small] + log1p(-x / 2 + x^2 / 6)
+  }
+  list(event = log_event, nonevent = -exp_eta)
```
[the remark this answers is not in the document]
Comment (choxos) [reply R/link_functions.R:137]: Correct, and I am recording it as a documented limit rather than fixing it here.

The value is genuinely lost upstream, not mis-transformed here. For this link `log(1 - p)` is `-exp(eta)`, which overflows to `-Inf` above `eta = 709.78`; `-exp(710)` has no double-precision representation, so by the time the generated quantity exists the magnitude is gone. Given `log_q = -Inf`, `Inf` is the correct transform of that input.

Recovering the finite link means the models emitting it as its own generated quantity rather than deriving it from `log_q`, which is the restructure you describe. Reachability is narrow: it needs a population-averaged probability whose complement is below 1e-308, so every integration point must sit above eta = 709.78. I have put the reason in the function so the next reader does not mistake it for a transformation bug, and left the change for a pull request that can carry the Stan-side quantity properly.

File: R/link_functions.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (choxos) [reply R/link_functions.R]: Fixed, same change as the Copilot comment on this line. Correction applies only at exactly 0 or 1; anything outside [0, 1] errors instead of being mapped to a boundary value; `NA` passes through. Regression test covers all four cases.
